src/kernels/helpers.ts

```diff
@@ -232,7 +232,12 @@ export function findPreferredKernel(
     );
 
     let preferredKernel: KernelConnectionMetadata | undefined = kernels[kernels.length - 1];
-    if (
+
+    if (!notebookMetadata || (preferredKernel && !isExactMatch(preferredKernel, notebookMetadata))) {
```

Comment:
As noted at standup. Instead of replacing compareKernels I'm just bolting a isExactMatch on top of the compareKernels result. Having spent a bunch of time digging around in compareKernels it's doing some smart stuff like making sure to prefer one type of launch over another, checking languages, ect...And given the work in #9615 I can see us doing things like returning a set of suggested kernels from sorted list if we don't have an exact match. Don't see a reason to keep compareKernels not doing its current job and just making sure that for now the top result fits our criteria.

Reply:
What if this was done at a higher level? (NotebookControllerManager I think).

That would allow all of the unit tests for findKernel to still pass. 

I think it might make sense to do the extra filtering at the top level because it might make other decisions other than 'exact' match later.

Just a thought. Not sure it matters.

Reply:
@rchiodo Combining your comment here with @DonJayamanne 's comment below. I think that I'll refactor this code path to be something more like rankKernels instead of findKernel. Mostly just a name change and instead of picking off the top element return back the entire sorted list. The current tests can be mostly the same if they just look at the top element.

The final consumer (ControllerManager) can use the exact match helper to figure out if it wants to use the top elements, and if there is no exact match can set multiple suggested items from the ranked list.